### DISCUSSION TITLE
Changing IP address of the DHCP interface

### DISCUSSION THREAD

**Discussion by intelliant01** [ltsp/ltsp#718] [Q&A]
When IP address is `192.168.67.1/24` in non-proxy mode, all works fine.
When I change it to `192.168.1.2/24` then I get the following errors -

```
root@server:~# ltsp dnsmasq -p0
RTNETLINK answers: Network is unreachable
LTSP command failed: ip -o route get 192.168.67.1
Could not parse output of: ip -o route get 192.168.67.1
RTNETLINK answers: Network is unreachable
LTSP command failed: ip -o route get 192.168.67.2
Could not parse output of: ip -o route get 192.168.67.2
Installed /usr/share/ltsp/server/dnsmasq/ltsp-dnsmasq.conf in /etc/dnsmasq.d/ltsp-dnsmasq.conf
Restarted dnsmasq

root@server:~# ltsp ipxe
RTNETLINK answers: Network is unreachable
LTSP command failed: ip -o route get 192.168.67.1
Could not parse output of: ip -o route get 192.168.67.1
RTNETLINK answers: Network is unreachable
LTSP command failed: ip -o route get 192.168.67.2
Could not parse output of: ip -o route get 192.168.67.2
Installed /usr/share/ltsp/server/ipxe/ltsp.ipxe in /srv/tftp/ltsp/ltsp.ipxe
Skipped existing /srv/tftp/ltsp/memtest.0
Skipped existing /srv/tftp/ltsp/memtest.efi
Skipped existing /srv/tftp/ltsp/snponly.efi
Skipped existing /srv/tftp/ltsp/undionly.kpxe

root@server:~# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: enp3s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 94:c6:91:a4:f9:b4 brd ff:ff:ff:ff:ff:ff
    inet 192.168.1.2/24 brd 192.168.1.255 scope global noprefixroute enp3s0
       valid_lft forever preferred_lft forever
3: wlp2s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default qlen 1000
    link/ether 38:ba:f8:89:b6:8d brd ff:ff:ff:ff:ff:ff

```
And obviously, the clients don't boot after this.

Why is it so? What could I be doing wrong? Or is it a bug?

**Comment by intelliant01** [comment]
However, if I manually change the following line in `ltsp-dnsmasq.conf` to ```dhcp-range=192.168.1.3,192.168.1.253,12h``` and restart dnsmasq, all works fine.

So something in `ltsp dnsmasq -p0`'s execution doesn't seem right.

**Comment by alkisg** [comment]
What is the output of these commands?

```shell
ip -o route get 192.168.67.1
ip r
```

**Comment by intelliant01** [comment]
When the interface IP is `192.168.1.2/24`, this is the output -

```
root@server:~# ip r
169.254.0.0/16 dev enp3s0 scope link metric 1000 
192.168.1.0/24 dev enp3s0 proto kernel scope link src 192.168.1.2 metric 100 
root@server:~# ip -o route get 192.168.67.1
RTNETLINK answers: Network is unreachable
root@server:~# ip -o route get 192.168.1.2
local 192.168.1.2 dev lo src 192.168.1.2 uid 0 \    cache <local> 
```

**Comment by intelliant01** [comment]
After a reboot, I don't get the `LTSP command failed:` errors that I have posted [herein](https://github.com/ltsp/ltsp/discussions/718#discussion-4098466) but neither is the correct ltsp-dnsmasq.conf generated and hence the clients still don't boot.

**Comment by alkisg** [comment]
The problem is that you don't have a default route.
Set up a route in network-manager or whatever else you're using, so that `ip r` shows a `default` line.

Additionally, make sure that your clients also get a default route from your DHCP server.

Btw is there a reason you want to re-implement this for 192.168.1.x instead of just using 192.168.67.x?

**Comment by intelliant01** [comment]
> The problem is that you don't have a default route.
Set up a route in network-manager or whatever else you're using, so that ip r shows a default line.

But neither do I have a default route set when IP is `192.168.67.1`. Why does it work fine in that case? Or do I? I shall be able to check this tomorrow and shall get back.

> Additionally, make sure that your clients also get a default route from your DHCP server.

This is non-proxy mode so the LTSP server is the DHCP server.

> Btw is there a reason you want to re-implement this for 192.168.1.x instead of just using 192.168.67.x?

Yes, the end users sometimes boot the same clients into Windows using embedded flash drives which have static addresses. And they want to simultaneously access the content hosted on the LTSP server while in Windows. They have other applications running on the network which are all configured with the 192.168.1.x subnet.

**Comment by alkisg** [comment]
LTSP needs a way to autodetect the internal adapter vs the Internet-facing one. That's what the `ip -o route get 192.168.67.1` does.
When the server is actually 192.168.67.1, a default gateway isn't needed, that command returns the correct adapter which points to itself.

> This is non-proxy mode so the LTSP server is the DHCP server.

And did you configure an option-router in your manual DHCP setup? You're reimplementing this stuff, so you need to replicate manually whatever LTSP does for 192.168.67.1.

**Comment by intelliant01** [comment]
When everything is working fine with 192.168.67.1 -

```
root@server:~#  ip r
169.254.0.0/16 dev enp3s0 scope link metric 1000 
192.168.67.0/24 dev enp3s0 proto kernel scope link src 192.168.67.1 metric 100 

root@server:~# ip -o route get 192.168.67.1
local 192.168.67.1 dev lo src 192.168.67.1 uid 0 \    cache <local> 
```

When IP address is changed to 192.168.1.2 via network manager *without* a default gateway -
```
root@server:~#  ip r
169.254.0.0/16 dev enp3s0 scope link metric 1000 
192.168.1.0/24 dev enp3s0 proto kernel scope link src 192.168.1.2 metric 100 

root@server:~# ip -o route get 192.168.1.2
local 192.168.1.2 dev lo src 192.168.1.2 uid 0 \    cache <local> 
```

When IP address is changed to 192.168.1.2 via network manager *with* a default gateway 192.168.1.1 -
```
root@server:~#  ip r
root@server:~# ip r
default via 192.168.1.1 dev enp3s0 proto static metric 100 
169.254.0.0/16 dev enp3s0 scope link metric 1000 
192.168.1.0/24 dev enp3s0 proto kernel scope link src 192.168.1.2 metric 100 

root@server:~# ip -o route get 192.168.1.2
local 192.168.1.2 dev lo src 192.168.1.2 uid 0 \    cache <local> 
```

But still it did not work. 

However eventually changing to `dhcp-range=192.168.1.20,192.168.1.250,12h` in the template at `/usr/share/ltsp/server/dnsmasq/ltsp-dnsmasq.conf` made it work without any errors.

So having the default gateway defined as `192.168.1.1` and the template change seems to work.

The IP change is being done by the user via Network Manager, so I think I shall use some automation via dispatcher to make the change in the template dynamically.

Do you see any issues with this logic that may conflict with the LTSP code? I ask as I see a lot of places in the LTSP code that 192.168.67.* is hard coded.

**Comment by alkisg** [comment]
> root@server:~# ip -o route get 192.168.1.2
> local 192.168.1.2 dev lo src 192.168.1.2 uid 0 \    cache <local> 

In that ^, we were interested for the output of `ip -o route get 192.168.67.1`, which is used by LTSP to detect the adapter to use. We're not interested for the route to 192.168.1.2.

> However eventually changing to dhcp-range=192.168.1.20,192.168.1.250,12h in the template at /usr/share/ltsp/server/dnsmasq/ltsp-dnsmasq.conf

You're supposed to create an /etc/dnsmasq.d/local.conf file with the dhcp-range that you want.
Then run `ltsp dnsmasq -p0 -r0`.

I.e. you're telling LTSP that you don't want to run its hardcoded real dhcp server at 192.168.67.x, and you don't want its proxy dhcp either.
And in local.conf you're telling dnsmasq that you want to run a real dhcp in the 192.168.1.x range.

**Comment by alkisg** [comment]
Another thing to note is, why are you using LTSP as a real dhcp server? If your clients are dual booted windows machines with hardcoded IPs etc, why don't you let your router do the dhcp, and tell dnsmasq do proxy-dhcp (i.e. the default ltsp setup)?
That way you wouldn't need any of the above.

**Comment by intelliant01** [comment]
>  Then run `ltsp dnsmasq -p0 -r0`

Thanks for the guidance regarding `r0`. I had completely missed this in the documentation or failed to understand it despite having read it all at least twice.

However, by just  using `local.conf`, everything is working perfectly fine without using `r0`. Then why use `r0`?

> Another thing to note is, why are you using LTSP as a real dhcp server? If your clients are dual booted windows machines with hardcoded IPs etc, why don't you let your router do the dhcp, and tell dnsmasq do proxy-dhcp (i.e. the default ltsp setup)?
That way you wouldn't need any of the above.

This is the million dollar question. I wanted to avoid explaining the complexity of the situation. But since you ask... :)

There are two situations that I have encountered recently :
One is a peculiar situation while deploying projects which are serving in low resource communities via a bus which is fitted with 20 clients and travels to schools in remote villages.

There is a webapp being served from an onboard Windows server which belongs to the `192.168.1.0/24`subnet and is beyond our control for IP configuration, etc. The clients when in Windows or Linux should be able to access them. If changing the Windows server IP to `192.168.67.0/24` subnet would have been possible, then things would have been simpler.
Also, when the clients boot into Windows with their static IPs, they should still be able to access the content being served from the LTSP server, hence in the most simplistic way the LTSP server should also be on the same subnet as the Windows clients. 

I was using a router 'A' to serve dhcp and LTSP server's dnsmasq was in proxy mode till a week ago. The router was connected via ethernet to a 4G SIM powered WiFi enabled router 'B' for internet access. Now their need evolved to be able to switch to the teacher's mobile phone hotspot if and when the router SIM's data plan is over mid-class or in situations where a recharge is not possible immediately. Hence using the router 'A' to route to the internet via 'B' over ethernet is of no use any more when they want dynamic switching to the hotspot. Thus, router 'A' became useless for serving IP + default gateway and the control must shift to the LTSP server.

Eventually, the life saver is a WiFi adapter on the LTSP server. So the LTSP clients can access internet routed through the WiFi adapter via the router B's WiFi or a teacher's mobile phone's WiFi hotspot. Have managed this with an `iptables...MASQUERADE` + `route del default dev <DHCPD_IFACE>` rule being dynamically fired whenever the WiFi connection comes up either via the WiFi router or the hotspot. I had tried with `NAT=1` but that seems to be designed only for dual NIC setups.

And there are several such buses. So needed a generic solution wherein I don't have to maintain customisations for each one of them and don't have to re-educate the on-field technical personnel not to change the static IP of the LTSP server ever. Now they may change the IPs via the Network Manager as on a standard Ubuntu system and yet LTSP will not break.

Another situation is where some "high-end" schools have so called networking policies that mandate a particular subnet but don't follow the best practices to serve IPs via a DHCP server or via MAC binding for lack of know-how or willingness to cooperate. Hence, the ability to set a static IP as per their liking may save me the hassle of such situations. But these is a fringe case as yet.

I am yet to figure out the significance of sticking to the hard coded `192.168.67.*` when a line in `local.conf` with just the `dhcp-range` still has everything working in both proxy and non-proxy modes.

Anyways, learnt a lot while working on this. Thanks as always for your prompt responses.
